File: wire-android-automation/common/src/main/java/com/wearezeta/auto/common/backend/models/Conversation.java
package com.wearezeta.auto.common.backend.models;

import org.json.JSONArray;
import org.json.JSONObject;

import java.util.ArrayList;
import java.util.List;
import java.util.Optional;

public class Conversation {
    private String id;
    private QualifiedID qualifiedID;
    private Optional<String> name = Optional.empty();
    /*
        RegularConv = 0 (= any group conversation, = any team conversation)
        SelfConv = 1
        One2OneConv = 2
        ConnectConv = 3
     */
    private Optional<Integer> type = Optional.empty();
    private Optional<String> teamId = Optional.empty();
    private String selfId;
    private List<QualifiedID> otherIds = new ArrayList<>();
    private String creatorId;
    private Integer receiptMode;
    private Integer messageTimerInMilliseconds;

    public Conversation() {
    }

    public String getId() {
        return id;
    }

    public void setId(String id) {
        this.id = id;
    }

    public void setQualifiedId(QualifiedID qualifiedID) {
        this.qualifiedID = qualifiedID;
    }

    public QualifiedID getQualifiedID() {
        return qualifiedID;
    }

    public Optional<String> getName() {
        return name;
    }

    public void setName(String name) {
        this.name = Optional.of(name);
    }

    public Optional<Integer> getType() {
        return type;
    }

    public void setType(int type) {
        this.type = Optional.of(type);
    }

    public Optional<String> getTeamId() {
        return teamId;
    }

    public void setTeamId(String teamId) {
        this.teamId = Optional.of(teamId);
    }

    public List<QualifiedID> getOtherIds() {
        return otherIds;
    }

    public void setOtherIds(List<QualifiedID> otherIds) {
        this.otherIds = new ArrayList<>(otherIds);
    }

    public String getCreatorId() {
        return creatorId;
    }

    public void setCreatorId(String creatorId) {
        this.creatorId = creatorId;
    }

    public String getSelfId() {
        return selfId;
    }

    public void setMessageTimerInMilliseconds(int timeoutInSeconds) {
        this.messageTimerInMilliseconds = timeoutInSeconds;
    }

    public int getMessageTimerInMilliseconds() {
        return messageTimerInMilliseconds;
    }

    public void setSelfId(String selfId) {
        this.selfId = selfId;
    }

    public boolean isReceiptModeEnabled() {
        return receiptMode != null && receiptMode > 0;
    }

    public void setReceiptMode(Integer receiptMode) {
        this.receiptMode = receiptMode;
    }

    public static Conversation fromJSON(JSONObject conversation) {
        final Conversation result = new Conversation();
        result.setId(conversation.getString("id"));

        result.setQualifiedId(QualifiedID.fromJSON(conversation.getJSONObject("qualified_id")));

        if (conversation.has("name") && !conversation.isNull("name")) {
            result.setName(conversation.getString("name").replaceAll("\uFFFC", "").trim());
        }

        if (!conversation.isNull("message_timer")) {
            result.setMessageTimerInMilliseconds(conversation.getInt("message_timer"));
        } else {
            result.setMessageTimerInMilliseconds(0);
        }

        if (conversation.has("type") && !conversation.isNull("type")) {
            result.setType(conversation.getInt("type"));
        }
        if (conversation.has("team") && !conversation.isNull("team")) {
            result.setTeamId(conversation.getString("team"));
        }

        if (conversation.has("qualified_id")) {
            JSONObject qualifiedID = conversation.getJSONObject("qualified_id");

            result.setQualifiedId(QualifiedID.fromJSON(qualifiedID));
        }

        // each 1:1 conversation is created with proteus, but has an "invisible" copy with mls protocol (if mls is enabled)
        // they have no creator, they simply exist between connected users
        // this is intended behavior
        if (conversation.isNull("creator")) {
            result.setCreatorId("null");
        } else {
            result.setCreatorId(conversation.getString("creator"));
        }

        result.setSelfId(conversation.getJSONObject("members").getJSONObject("self").getString("id"));

        final JSONArray others = conversation.getJSONObject("members").getJSONArray("others");

        List<QualifiedID> otherIds = new ArrayList<>();
        for (int i = 0; i < others.length(); ++i) {
            QualifiedID othersQualifiedID = QualifiedID.fromJSON(others.getJSONObject(i).getJSONObject("qualified_id"));
            otherIds.add(othersQualifiedID);
        }
        result.setOtherIds(otherIds);

        if (conversation.has("receipt_mode") && !conversation.isNull("receipt_mode")) {
            result.setReceiptMode((conversation.getInt("receipt_mode")));
        }
        return result;
    }
}

File: tests/testsSupport/src/main/service/models/Conversation.kt
/*
 * Wire
 * Copyright (C) 2025 Wire Swiss GmbH
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program. If not, see http://www.gnu.org/licenses/.
 */
package service.models


import org.json.JSONArray
import org.json.JSONObject

data class Conversation(
    var id: String = "",
    var qualifiedID: QualifiedID = QualifiedID(),
    var name: String? = null,
    /*
        RegularConv = 0 (= any group conversation, = any team conversation)
        SelfConv = 1
        One2OneConv = 2
        ConnectConv = 3
     */
    var type: Int? = null,
    var teamId: String? = null,
    var selfId: String = "",
    var otherIds: List<QualifiedID> = emptyList(),
    var creatorId: String = "",
    var receiptMode: Int? = null,
    var protocol: String = "",
    var messageTimerInMilliseconds: Int = 0
) {
    val isReceiptModeEnabled: Boolean
        get() = receiptMode?.let { it > 0 } ?: false

    companion object {
        fun fromJSON(json: JSONObject): Conversation {
            val conversation = Conversation()

            conversation.id = json.getString("id")
            conversation.protocol = json.getString("protocol")
            conversation.qualifiedID = QualifiedID.fromJSON(json.getJSONObject("qualified_id"))

            conversation.name = json.optString("name", null)?.takeIf { it.isNotBlank() }
                ?.replace("\uFFFC", "")?.trim()

            conversation.messageTimerInMilliseconds = json.optInt("message_timer", 0)
            conversation.type = json.optInt("type").takeIf { json.has("type") && !json.isNull("type") }
            conversation.teamId = json.optString("team", null).takeIf { json.has("team") && !json.isNull("team") }

            conversation.creatorId = if (json.isNull("creator")) "null" else json.getString("creator")

            val members = json.getJSONObject("members")
            conversation.selfId = members.getJSONObject("self").getString("id")

            val otherArray: JSONArray = members.getJSONArray("others")
            conversation.otherIds = (0 until otherArray.length()).map { i ->
                val qualified = otherArray.getJSONObject(i).getJSONObject("qualified_id")
                QualifiedID.fromJSON(qualified)
            }

            if (json.has("receipt_mode") && !json.isNull("receipt_mode")) {
                conversation.receiptMode = json.getInt("receipt_mode")
            }

            return conversation
        }
    }
}
